src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleCloudStorageDAO.scala

```diff
@@ -49,38 +55,69 @@ class HttpGoogleCloudStorageDAO(
+            blocking {
+              inserter.execute
+            }
+          }
+        }
+      }
+    }
+
+    def insertOwnerMember: (Seq[Group]) => Future[Member] = { _ =>
+      retry(when500) {
+        () => Future {
+          val ownersGroupId = toGroupId(bucketName, WorkspaceAccessLevel.Owner)
           val owner = new Member().setEmail(userInfo.userEmail).setRole(groupMemberRole)
-          val ownerInserter = directory.members.insert(ownersGroupId,owner)
-          retry(ownerInserter.execute,when500)
-
-          val readersEntity = makeGroupEntityString(toGroupId(bucketName,WorkspaceAccessLevel.Read))
-          val writersEntity = makeGroupEntityString(toGroupId(bucketName,WorkspaceAccessLevel.Write))
-          val ownersEntity = makeGroupEntityString(ownersGroupId)
-          val bucket = new Bucket().setName(bucketName).setAcl(List(
-              newBucketAccessControl(readersEntity,WorkspaceAccessLevel.Read),
-              newBucketAccessControl(writersEntity,WorkspaceAccessLevel.Write),
-              newBucketAccessControl(ownersEntity,WorkspaceAccessLevel.Owner)
-            )).setDefaultObjectAcl(List(
-              newObjectAccessControl(readersEntity,WorkspaceAccessLevel.Read),
-              // NB: writers have read access to objects -- there is no write access, objects are immutable
-              newObjectAccessControl(writersEntity,WorkspaceAccessLevel.Read/*sic!*/),
-              newObjectAccessControl(ownersEntity,WorkspaceAccessLevel.Owner)
-            ))
-          val inserter = getStorage(getBucketCredential(userInfo)).buckets.insert(projectId,bucket)
-          retry(inserter.execute,when500)
+          val ownerInserter = directory.members.insert(ownersGroupId, owner)
+          blocking {
+            ownerInserter.execute
+          }
+        }
+      }
+    }
+
+    def insertBucket: (Member) => Future[Unit] = { _ =>
+      retry(when500) {
+        () => Future {
+          val bucketAcls = groupAccessLevelsAscending.map(ac => newBucketAccessControl(makeGroupEntityString(toGroupId(bucketName, ac)), ac))
+          val defaultObjectAcls = groupAccessLevelsAscending.map(ac => {
+            // NB: writers have read access to objects -- there is no write access, objects are immutable
+            newObjectAccessControl(makeGroupEntityString(toGroupId(bucketName, ac)), if (ac == WorkspaceAccessLevel.Owner) ac else WorkspaceAccessLevel.Read)
+          })
+
+          val bucket = new Bucket().setName(bucketName).setAcl(bucketAcls).setDefaultObjectAcl(defaultObjectAcls)
+          val inserter = getStorage(getBucketCredential(userInfo)).buckets.insert(projectId, bucket)
+          blocking {
+            inserter.execute
+          }
         }
       }
     }
+
+    val doItAll = insertGroups(workspaceName, bucketName) flatMap insertOwnerMember flatMap insertBucket
+    doItAll.transform(f => f, rollbackGroups)
```

Comment:
I believe that there's a subtle race condition here relating to the use of traverse.  (It isn't really important in this case, since the consequence will only be to abandon an unused google group, but since this is intended to be best methods demo code, I feel the need to point it out.)  Traverse will put 3 jobs -- one to create each group -- into the pool where they will be acted upon as the execution context has resources.  However, what traverse returns is a Future[Seq[Group]], and that Future gets completed upon <em>the first detected failure</em>.  It might thus be possible that the recovery action (rollbackGroups) might be invoked before the lagging create group jobs even have a chance to run.  When the job is finally scheduled, it would then create a group that is simply abandoned.
I'm not certain what the most elegant solution is, but I think you might need to create a Future that completes only when <em>all</em> the parallel jobs complete.  You could call it join(Seq[Future[T]]).